nautobot/ipam/views.py

```diff
@@ -1401,23 +1389,19 @@ def alter_obj(self, obj, request, url_args, url_kwargs):
         return obj
 
 
-class ServiceBulkImportView(generic.BulkImportView):  # 3.0 TODO: remove, unused
+class ServiceUIViewSet(NautobotUIViewSet):  # 3.0 TODO: remove, unused BulkImportView
+    model = Service
+    bulk_update_form_class = forms.ServiceBulkEditForm
+    filterset_class = filters.ServiceFilterSet
+    filterset_form_class = forms.ServiceFilterForm
+    form_class = forms.ServiceForm
     queryset = Service.objects.all()
-    table = tables.ServiceTable
-
-
-class ServiceDeleteView(generic.ObjectDeleteView):
-    queryset = Service.objects.all()
-
-
-class ServiceBulkEditView(generic.BulkEditView):
-    queryset = Service.objects.select_related("device", "virtual_machine")
-    filterset = filters.ServiceFilterSet
-    table = tables.ServiceTable
-    form = forms.ServiceBulkEditForm
-
-
-class ServiceBulkDeleteView(generic.BulkDeleteView):
-    queryset = Service.objects.select_related("device", "virtual_machine")
-    filterset = filters.ServiceFilterSet
-    table = tables.ServiceTable
+    serializer_class = serializers.ServiceSerializer
+    table_class = tables.ServiceTable
+
+    def get_queryset(self):
+        if self.action in ["bulk_update", "bulk_destroy"]:
+            self.queryset = Service.objects.select_related("device", "virtual_machine")
+        elif self.action == "retrieve":
+            self.queryset = Service.objects.prefetch_related("ip_addresses")
+        return super().get_queryset()
```

Comment:
Not sure the conditional logic is needed here - both the retrieve view and the ServiceTable display all three of the specified fields, so I'd think just declaring `queryset = Service.objects.select_related("device", "virtual_machine").prefetch_related("ip_addresses")` above would be sufficient.

Reply:
okay, so there is no place where we are not getting this data, so this would actually optimize those other queries.